### PR TITLE
Port component/command diff storage/deserializer generators

### PR DESCRIPTION
**Pull request by paulbalaji**
#### Description

ported

- component diff storage generator
- component diff deserializer generator
- command diff storage generator
- command diff deserializer generator

#### Tests

see generated code

#### Documentation

will add changelog entry in final core port pr

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffDeserializerGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                    });
+                });
+            }).Format();
+        }
+
+        private static Text GenerateDiffCommandDeserializer(UnityCommandDetails command, string qualifiedNamespace, string componentName)
+        {
+            Logger.Trace($"Generating {qualifiedNamespace}.{componentName}.{command.CommandName}DiffCommandDeserializer class.");
+
+            var receivedRequestType = $"{command.CommandName}.ReceivedRequest";
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffDeserializerGenerator.cs]: These can probably go away

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffStorageGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    }}
+}}
+");
+        }
+
+        private static Text GenerateCommandsToSendStorage(UnityCommandDetails command, string qualifiedNamespace, string componentName)
+        {
+            Logger.Trace($"Generating {qualifiedNamespace}.{componentName}.{command.CommandName}CommandsToSendStorage class.");
+
+            var requestType = $"{command.CommandName}.Request";
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffStorageGenerator.cs]: These can also go?

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffDeserializerGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                                "diff.AddComponentUpdate(update, op.EntityId, op.Update.ComponentId, updateId);"
+                            });
+
+                        if (eventDetailsList.Count > 0)
+                        {
+                            m.Line("var eventsObject = op.Update.SchemaData.Value.GetEvents();");
+
+                            foreach (var ev in eventDetailsList)
+                            {
+                                var eventType = $"{ev.EventName}.Event";
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffDeserializerGenerator.cs]: We can probably inline this?

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffDeserializerGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    var schemaUpdate = SchemaComponentUpdate.Create();
+    var componentUpdate = new ComponentUpdate(ComponentId, schemaUpdate);
+    Serialization.SerializeUpdate(update.Update, schemaUpdate);
+    serializedMessages.AddComponentUpdate(componentUpdate, update.EntityId.Id);
+}
+");
+
+                        foreach (var ev in eventDetailsList)
+                        {
+                            var eventType = $"{ev.EventName}.Event";
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffDeserializerGenerator.cs]: inline?

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                    ns.Type($"public partial class {componentDetails.ComponentName}", partial =>
+                    {
+                        Logger.Trace($"Generating {qualifiedNamespace}.{componentDetails.ComponentName}.DiffComponentStorage class.");
+
+                        var classDefinition = new StringBuilder("public class DiffComponentStorage : IDiffUpdateStorage<Update>, IDiffComponentAddedStorage<Update>, IDiffAuthorityStorage");
+
+                        foreach (var ev in eventDetailsList)
+                        {
+                            var eventType = $"{ev.EventName}.Event";
+                            classDefinition.Append($"{Environment.NewLine}    , IDiffEventStorage<{eventType}>");
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs]: This should be using the indenting helper

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+private MessageList<ComponentUpdateReceived<Update>> updateStorage =
+    new MessageList<ComponentUpdateReceived<Update>>();
+
+private MessageList<AuthorityChangeReceived> authorityChanges =
+    new MessageList<AuthorityChangeReceived>();
+");
+
+                            foreach (var ev in eventDetailsList)
+                            {
+                                var eventType = $"{ev.EventName}.Event";
```
Comment (zeroZshadow) [inline workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs]: I've seen this so often now, it should probably be a property on the `ev` type

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffStorageGenerator.cs (hunk 2 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/CommandDiffStorageGenerator.cs]: I've left it in where we use it several times

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs (hunk 6 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs]: indeed - it should be :)

File: workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs (hunk 5 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/ComponentDiffStorageGenerator.cs]: disagree - as this is a lot clearer to read

**Pull request merged**